### ISSUE TITLE
NormalizedArrays.Arrays.CommaAfterLast sniff is wrong for Moodle

### ISSUE THREAD

**Issue by timhunt** [moodlehq/moodle-cs#76]
The newly added NormalizedArrays.Arrays.CommaAfterLast sniff gtom #60 does not correctly impleement Moodle coding style.

The good example

```php
$myarray = [1, 2, 3, 'Stuff', 'Here',
    $a, $b, $c, 56.44, $d, 500];
```

from https://moodledev.io/general/development/policies/codingstyle#numerically-indexed-arrays now gives a spurious failure.

**Comment by paulholden** [comment]
It's also tripping up on this, which I think isn't correct and possibly related to Tim's example above:

```
            'Filter cohort name' => ['cohort:name', [
                'cohort:name_operator' => text::IS_EQUAL_TO,
                'cohort:name_value' => 'My cohort',
            ], true],
```

It suggests the following odd fix:

```
            'Filter cohort name' => ['cohort:name', [
                'cohort:name_operator' => text::IS_EQUAL_TO,
                'cohort:name_value' => 'My cohort',
            ], true, ], // Odd because the outer array isn't multiline, but the inner array is
```

**Comment by stronk7** [comment]
Just for the records, I've tried with the `Squiz.Arrays.ArrayDeclaration` Sniff instead of the `NormalizedArrays.Arrays.CommaAfterLast` and it faces the same problem.

Surely all this is because various of the standards (normal ones) out there always ask for multi-value arrays to have the square brackets (opening and close) in different lines...

In any case, the Squiz one does the same and also enforces way more rules, so it's not a solution for this case. Just sharing to avoid anybody trying the same...

About @paulholden example above... I must confess that a "personal* rule that I try to follow always is to avoid more than one square bracket in the same line (unless the inner one fits in one line). So, surely, for me, it should be:

```
            'Filter cohort name' => [
                'cohort:name', [
                    'cohort:name_operator' => text::IS_EQUAL_TO,
                    'cohort:name_value' => 'My cohort',
                ],
                true,
            ], // Odd because the outer array isn't multiline, but the inner array is
```

Or:

```
            'Filter cohort name' => [ 'cohort:name',
                [
                    'cohort:name_operator' => text::IS_EQUAL_TO,
                    'cohort:name_value' => 'My cohort',
                ], true
            ], // Odd because the outer array isn't multiline, but the inner array is
```

In any case, that's just a personal preference... heh. Only commenting. Note the first one has comma after `true`, but the second does not.

So, given all the examples... I think that the detail that the sniff is missing is to be able to detect how many items there are in the line, if there are more that one (or said with other words, if the previous item ended in the same line.... then the comma should not be required).

Ciao :-)

**Comment by stronk7** [comment]
This seems to be working here:

```
diff --git a/NormalizedArrays/Sniffs/Arrays/CommaAfterLastSniff.php b/NormalizedArrays/Sniffs/Arrays/CommaAfterLastSniff.php
index d03d1ff..6472e5c 100644
--- a/NormalizedArrays/Sniffs/Arrays/CommaAfterLastSniff.php
+++ b/NormalizedArrays/Sniffs/Arrays/CommaAfterLastSniff.php
@@ -159,6 +159,12 @@ final class CommaAfterLastSniff implements Sniff
                     return;
                 }
 
+                // If the line of the last non-empty token is the same as the closer, we're not
+                // going to require a comma. TODO: Put this under some Sniff configuration setting.
+                if ($tokens[$lastNonEmpty]['line'] === $tokens[$closer]['line']) {
+                    return;
+                }
+
                 $error     = 'There should be a comma after the last array item in a %s array.';
                 $errorCode = 'Missing' . $errorCode;
                 $data      = [$phrase];
```
(basically, if the array closer is in the same line than the last element of the array... then disable the comma requirement).

I've done some quick (and incomplete) testing and it seems to be working ok for the cases commented in this issue.

![quick_and_incomplete_ests](https://github.com/moodlehq/moodle-cs/assets/167147/32266e1f-0fe8-4b18-87bb-e79ada047e53)

**Comment by timhunt** [comment]
Those examples look promising to me.

**Comment by stronk7** [comment]
I've created https://github.com/PHPCSStandards/PHPCSExtra/issues/283 upstream, let's see.

**Comment by stronk7** [comment]
It seems that the upstream problem has been fixed, so I'm going to add a few own tests here and, as soon as there is a new release of PHPCSExtra, I'll incorporate it to us.

Ciao :-)

**Comment by stronk7** [comment]
For the records, these are the tests I'm going to add, and note that, with the still not released, now all the cases in the fixtures file are passing ok. If you've any other case worth testing, let me know.

![current_vs_new](https://github.com/moodlehq/moodle-cs/assets/167147/4743d109-f016-4e3f-a663-4eda9e0831c2)

**Comment by stronk7** [comment]
As far as it seems that the PHPCSExtra 1.2.0 release is getting long... and this is hitting us a lot, I'm planning to "pick" the current Sniff and add it to our own Sniffs.

Of course, in the future, once PHPCSExtra 1.2.0 is released... we'll switch to it (will create issue about that).

What do you think? @andrewnicols , @timhunt ...

**Comment by stronk7** [comment]
Ok, for the record, after a little bit of internal discussion we have decided to:

1. Go ahead with #81 that will fix this issue.
2. Make a moodle-cs release with the fix and spread it to dependent tools (CiBoT, codechecker, moodle-plugin-ci).
3. Once PHPCSExtra 1.2.0 is released, in #82 we'll revert the commit and will switch to upstream sniff.

On it!

**Comment by stronk7** [comment]
Ok, [v3.3.11](https://github.com/moodlehq/moodle-cs/releases/tag/v3.3.11) has been released and that includes the fix for that. CiBoT will catch up in minutes.

Later this week I'll release codechecker and moodle-plugin-ci with updated dependencies.

And, at some time in the future, with #82 , we'll be back to use the upstream Sniff.

That's all, ciao :-)
